docs/proposals/2022-07-13-appset-progressive-rollout-strategy.md

```diff
@@ -0,0 +1,165 @@
+              values:
+                - qa
+        - matchExpressions:
+            - key: env
+              operator: In
+              values:
+                - us-east-2
+                - eu-west-1
+                - ap-southeast-1
+          maxUpdate: 1 # maxUpdate supports both integer and percentage string values
+  template:
+    metadata:
+      name: '{{cluster}}-guestbook'
+    spec:
+      source:
+        repoURL: https://github.com/infra-team/cluster-deployments.git
+        targetRevision: HEAD
+        path: guestbook/{{cluster}}
+      destination:
+        server: '{{url}}'
+        namespace: guestbook
+```
+
+In the above example, when the guestbook ApplicationSet is created or modified, the Application resources are each updated in the order defined in `strategy.rollingUpdate`. In this case, all Applications whose desired resource state matches the label expression `env: dev` are updated to match the template, as the `maxUpdate` is set to zero. The rolling update strategy progresses after the first set of Applications has successfully completed. Progress towards the next step starts only after the current step has completely finished, regardless of the `maxUpdate` value. The `maxUpdate` field only throttles the total number of matching Applications updating in the current step. After the first step completes, the ApplicationSet updates all Application resources with label `env: qa` at the same time, because `maxUpdate` is undefined. Finally, during the third step, the Application resources labeled `env: us-east-2`, `env: eu-west-1`, or `env: ap-southeast-1` are updated, one by one, as the `maxUpdate` for the final step is 1.
+
+An Application rollout is considered “complete” when the Application resource has been:
+- Synced successfully.
+- Moved into a “Progressing” state.
+- Moved out of a “Progressing” state and into a “Healthy” state.
+
+
+#### Use case 2:
+As a user, I would like to continue to use the current simultaneous Application update behavior of the ApplicationSet controller.
+
+If no strategy is provided, we propose defaulting to an `AllAtOnce` strategy, which maintains the current behavior.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Initial ApplicationSet Creation
+Application resource creation from an ApplicationSet with a defined strategy looks much like the update process. When a brand new ApplicationSet is first created with a rollout strategy specified, the desired Application resource metadata labels are used to determine when each Application resource is created. Each Application created will be created in the order defined by the steps, if any, and advance to the next step only when a step completes successfully. The same applies if an ApplicationSet is modified to target a different set of destination clusters or namespaces, Applications are created or updated in the order defined by their desired state and the defined step order in the strategy.
```

Comment:
> The same applies if an ApplicationSet is modified to target a different set of destination clusters or namespaces, Applications are created or updated in the order defined by their desired state and the defined step order in the strategy.

What if not all Applications are updated? Suppose the destination namespace field is templated - the source of truth for a git generator might update an Application in step 2 but none in step 1. Given the Sync->Progressing->Healthy checklist above, it's not clear that step 1 would ever be considered "complete," and we might therefore never reach step 2.

Will the rolling update strategy be respected for non-destination Application spec changes? For example, if I make a change that affects the `spec.template.source.helm.parameters[0].value` field, will that update be pushed immediately, or will the full strategy be followed again?

Reply:
Just tested this scenario with two Applications, one in step 1, and one in step 2. I made a change that would only be applied to Application 2.

```
status:
  applicationStatus:
  - application: hello-world-nginx-awscmhdev
    lastTransitionTime: "2022-08-11T22:06:47Z"
    message: Application moved to Pending status, watching for the Application resource
      to start Progressing.
    observedGeneration: 7
    status: Pending
  - application: hello-world-nginx-awscmhqa2
    lastTransitionTime: "2022-08-11T22:06:47Z"
    message: Application is out of date with the current AppSet generation, setting
      status to Waiting.
    observedGeneration: 6
    status: Waiting
```

Currently, the controller does expect that all tracked Applications are going to be updated, and waits for Application 1 to enter the Progressing state (it never does). This eventually hits a hardcoded 5 minute timeout which automatically bumps an Application in a Pending state to Healthy if the Application's status is Healthy.

So in summary, it will currently spend 5 minutes on each step that contains an Application with no changes before timing out and proceeding to the next step. I believe I can address this limitation by tracking a hash of the generated Application spec, and skipping straight to the Healthy state for each generated Application that has no changes.

Reply:
I also tested the case where an external change (in my case a cluster secret label) changed the generated Application spec without changing the ApplicationSet spec. This case is also not handled well, but should be covered if I switch to comparing the individual hashes of each generated Application instead of just the ApplicationSet's `resourceVersion`.

Reply:
The edge cases raised are now handled in https://github.com/argoproj/argo-cd/pull/10048.